lib/converters/ast-to-finite-field.js

```diff
@@ -0,0 +1,202 @@
+  } 
+
+  return [ m[0][0], m[1][0] ];
+}
+
+function finite_field_evaluate_ast(tree, bindings, modulus, level) {
+  if (typeof tree === 'string') {
+    if (tree === "e") {
+      return new ZmodN([numberTheory.primitiveRoot(modulus)], modulus);
+    }
+
+    if (tree === "pi") {
+      if (modulus % 2 == 0)
+        return new ZmodN([modulus / 2], modulus);
+      else
+        // Probably a really bad idea
+        return new ZmodN([NaN], modulus);
+    }
+
+    if (tree === "i")
+      return new ZmodN([0], 1);
+
+    if (tree in bindings)
+      return new ZmodN([bindings[tree]], modulus);
+
+    return tree;
+  }
+
+  if (typeof tree === 'number') {
+    if (Number.isInteger( tree )) {
+      return new ZmodN([tree], modulus);
+    } else {
+      let r = rationalApproximation( tree );
+      let numerator = new ZmodN([r[0]], modulus);
+      let denominator = new ZmodN([r[1]], modulus);
+      return numerator.divide( denominator );
+    }
+  }
+
+  var operator = tree[0];
+  var operands = tree.slice(1);
+  
+  if (operator === "apply") {
+    if (typeof operands[0] !== 'string')
+      throw Error("Non string functions not implemented for finite fields");
+
+    if (operands[0] === "exp") {
+      var base = finite_field_evaluate_ast("e", {}, modulus, level);
+      var exponent = finite_field_evaluate_ast(operands[1], bindings, numberTheory.eulerPhi(modulus), level + 1);
+      return base.power(exponent);
+    }
+
+    if (operands[0] === "sec") {
+      return finite_field_evaluate_ast(['/', 1, ['cos', operands[1]]], bindings, modulus, level);
+    }
+
+    if (operands[0] === "tan") {
+      var phi = numberTheory.eulerPhi(modulus);
+      var x = finite_field_evaluate_ast(operands[1], bindings, phi, level + 1);
+      return finite_field_evaluate_ast(['/', ['sin', x], ['cos', x]], bindings, modulus, level);
+    }
+
+    if (operands[0] === "sin") {
+      var root = numberTheory.primitiveRoot(modulus);
+      var g = new ZmodN([root], modulus);
+      var phi = numberTheory.eulerPhi(modulus);
+      if (phi % 4 != 0)
+        return new ZmodN([NaN], modulus);
+
+      var i = new ZmodN([numberTheory.powerMod(root, phi / 4, modulus)], modulus);
+      var x = finite_field_evaluate_ast(operands[1], bindings, phi, level + 1);
+      return (g.power(x).subtract(g.power(x.negate()))).divide(i.add(i));
+    }
+
+    if (operands[0] === "cos") {
+      var g = new ZmodN([numberTheory.primitiveRoot(modulus)], modulus);
+      var x = finite_field_evaluate_ast(operands[1], bindings, numberTheory.eulerPhi(modulus), level + 1);
+      return (g.power(x).add(g.power(x.negate()))).divide(new ZmodN([2], modulus));
+    }
+
+    if (operands[0] === "log") {
+      return new ZmodN([NaN], NaN);
+    }
+  }
+      
+  if (operator == "^") {
+    var base = finite_field_evaluate_ast(operands[0], bindings, modulus, level);
+    var exponent = finite_field_evaluate_ast(operands[1], bindings, numberTheory.eulerPhi(modulus), level + 1);
```

Comment:
The calculation of the exponent fails for `x^0.5`, as the expression for the exponent ends up having no values.